doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,457 @@
+
+
+Partition and Reactor thread
+----------------------------
+In early stage development, PoseidonStore will employ static allocation of partition. The number of sharded partitions
+is fixed and the size of each partition also should be configured before running cluster.
+But, the number of partitions can grow as below. We leave this as a future work.
+   +------+------+-------------+----------------+
+   | SP 1 | SP N | -->     <-- | gobal partition|
+   +------+------+-------------+----------------+ 
+Also, each reactor thread has a static set of SPs.
+
+
+Cache
+-----
+There are mainly two cache data structures; onode cache and block cache.
+It looks like below.
+
+1.Onode cache:
+lru_map <OID, OnodeRef>;
+Onode {
+  extent tree block_maps;
+  clone tree clones;
+  omap lists omaps;
+}
+2. Block cache:
+Omap cache --> lru_map <OID, <omap_key, value>>
+Data cache --> lru_map <OID, <extent, value>>
+
+To fill the onode data structure, the target onode needs to be retrieved using the prefix tree.
+Block cache is used for caching a block contents. For a transaction, all the updates to blocks 
+(including object meta block, data block) are first performed in the in-memory block cache.
+After writing a transaction to the WAL, the dirty blocks are flushed to their respective locations in the 
+respective partitions.
+PoseidonStore can configure cache size for each type. Simple LRU cache eviction strategy can be used for both.
+
+
+Sharded partitions (with cross-SP transaction)
+--------------------------------------------
+Entire disk space is divided into a number of chunks called sharded partitions (SP).
+The prefixes of the parent collection ID (original collection ID before collection splitting. That is, hobject.hash) 
+is hashed to map any collections to SPs.
+We can use BlueStore's approach for collection splitting, changing the number of significant bits for the collection prefixes.
+Because the prefixes of the parent collection ID do not change even after collection splitting, the mapping between 
+the collection and SP is maintained.
+The number of SPs may be configured to match the number of CPUs allocated for each disk so that each SP can hold 
+a number of objects large enough for cross-SP transaction not to occur.
+
+In case of need of cross-SP transaction, we could use the global WAL. The coodinator thread (mainly manages global partition) handles 
+cross-SP transaction via acquire the source SP and target SP locks before processing the cross-SP transaction. 
+Source and target probably are blocked.
+
+For the load unbalanced situation, 
+Poseidonstore can create partitions to make full use of entire space efficiently and provide load balaning.
+
+
+COW/Clone
+---------
+As for CoW/Clone, original version of the tree is stored in clones[0] in the onode. 
+Only changes from clones[0] will be recorded the next entry of clones.
+Note that clones[2] will contain entire changes in clone[1].
```

Comment:
I need more information here.

Reply:
As for CoW/Clone, a clone has an onode like other normal objects.
As shown the figure as below, by using shared_blob tracked by shared_blob_list in the superblock,
the blocks for the clone are managed using shared_block with reference count.


    +---------+        +---------+
    | Onode A |---     | Clone A |
    +---------+  |     +---------+
        v       ---v      v
    +--------------+----------------------------+
    | Data block 1 | Data block 2 (shared_blob) |
    +--------------+----------------------------+

Reply:
I don't really understand what the clones element of an onode means.  By "shared_blob_list" in the superblock, what do you mean?

Reply:
Hm... Isn't this similar to existing Ceph or btrfs?
Assuming that there is an object A,
clones from the object A will have an independent OID like xxx:0 and xxx:1 (unlike xxx:head).
So, these clones will have its own onode.

Although each clone has its own onode, data blocks should be shared between the original object and clones
if there are no changes on them to minimize the space overhead.
To do so, the reference count for the data blocks is needed to manage those shared data blocks.

To deal with the data blocks which has the reference count, the practical way is to use the shared block information recorded in the global area (within a sharded partition) like superblock. 
shared_blob_list means shared block information for the data blocks.

Reply:
I guess I don't understand specifically why each onode needs to in some sense track clones.  I expect that some extent indirections necessarily point at an indirection capable of handling ref counting (the shared_blob structure), but I don't understand yet how the shard_blob works or what the onode 'clones' member does.

Reply:
I removed 'clones' in the onode, there is no need to have such member variable. 
Like current Ceph, Extent has a Blob, and the Blob can have shared_blob if any.